File: bytedance/199.py
# 199. 二叉树的右视图
# 1. BFS：每层遍历的最右边的存入
# 2. DFS：根-右-左，可保证每层第一次访问的是右边的

from typing import List
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def rightSideView(self, root: TreeNode) -> List[int]:
        res = []

        def dfs(root, depth):
            if not root:
                return
            if len(res) <= depth:
                res.append(0)
            res[depth] = root.val # 利用深度判断第一次进来的肯定是右边的
            dfs(root.left, depth+1)
            dfs(root.right, depth+1)

        dfs(root, 0)
        
        return res

    def ightSideView2(self, root: TreeNode) -> List[int]:
        if not root : return []
        res = []

        def bfs(root):
            queue = [root]
            while queue:
                nxt = []
                res.append(queue[-1].val)
                for node in queue:
                    if node.left:
                        nxt.append(node.left)
                    if node.right:
                        nxt.append(node.right)
                queue = nxt
        
        bfs(root)
        return res

File: bytedance/dy/199.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

// 右侧
// BFS：每一层的最后一个节点
// DFS：加入辅助depth，首次size==depth时add，右视图先右再左
class Solution {

    // BFS
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> ans = new ArrayList<>();
        if (root == null) return ans;

        LinkedList<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            int size = queue.size();
            // 加入一层
            for (int i = 0; i < size; i++) {
                TreeNode node = queue.poll();
                if (node.left != null) queue.offer(node.left);
                if (node.right != null) queue.offer(node.right);

                if (i == size - 1) ans.add(node.val);
            }
        }

        return ans;
    }


    // DFS
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> ans = new ArrayList<>();
        dfs(root, 0, ans);
        return ans;
    }

    private void dfs(TreeNode root, int depth, List<Integer> ans) {
        if (root == null) return;

        // 相等的时候add
        if (ans.size() == depth) {
            ans.add(root.val);
        }

        depth++;
        dfs(root.left, depth, ans);
        dfs(root.right, depth, ans);
    }



    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) { val = x; }
  }
}